File: app/src/utils/deficiency-profile.ts
import type { NutrientKey, NutrientFruit } from '../types';
import { NUTRIENT_META } from './nutrition-meta';

export type BiologicalSex = 'male' | 'female';
export type AgeRange = '19-30' | '31-50' | '51+';
export type DietPattern = 'omnivore' | 'pescatarian' | 'vegetarian' | 'vegan';
export type HealthFocus = 'heart' | 'bone' | 'energy' | 'gut' | 'immune';
export type PregnancyStatus = 'pregnant' | 'breastfeeding';
export type LifestyleFactor = 'smoker' | 'alcohol' | 'caffeine';
export type Symptom = 'fatigue' | 'cramps' | 'bruising' | 'colds';
export type FamilyCondition = 'osteoporosis' | 'heart_disease' | 'neurodegeneration' | 'diabetes';

export interface WizardAnswers {
  sex: BiologicalSex;
  ageRange: AgeRange;
  dietPattern: DietPattern;
  healthFocus: HealthFocus[];
  pregnancyStatus: PregnancyStatus | null;
  lifestyleFactors: LifestyleFactor[];
  symptoms: Symptom[];
  familyHistory: FamilyCondition[];
}

export interface ScoredFood {
  food: NutrientFruit;
  score: number;
  topNutrients: { key: NutrientKey; label: string; percentDV: number }[];
}

type DeficiencyWeights = Map<NutrientKey, number>;

const BASELINE_WEIGHTS: Partial<Record<NutrientKey, number>> = {
  fiber_g: 0.3,
  potassium_mg: 0.3,
  magnesium_mg: 0.25,
  vitamin_e_mg: 0.2,
  vitamin_c_mg: 0.15,
  calcium_mg: 0.2,
  vitamin_a_mcg: 0.15,
};

const DIET_PATTERN_WEIGHTS: Record<DietPattern, Partial<Record<NutrientKey, number>>> = {
  omnivore: {},
  pescatarian: {
    iron_mg: 0.15,
    zinc_mg: 0.1,
  },
  vegetarian: {
    vitamin_b12_mcg: 0.4,
    iron_mg: 0.3,
    zinc_mg: 0.2,
    calcium_mg: 0.15,
    protein_g: 0.15,
    selenium_mcg: 0.1,
  },
  vegan: {
    vitamin_b12_mcg: 0.6,
    iron_mg: 0.35,
    zinc_mg: 0.3,
    calcium_mg: 0.3,
    protein_g: 0.25,
    selenium_mcg: 0.2,
    vitamin_b2_mg: 0.15,
  },
};

const SEX_AGE_WEIGHTS: Record<string, Partial<Record<NutrientKey, number>>> = {
  'female:19-30': { iron_mg: 0.3, vitamin_b9_mcg: 0.2 },
  'female:31-50': { iron_mg: 0.3, vitamin_b9_mcg: 0.2 },
  'female:51+': { vitamin_b12_mcg: 0.2, calcium_mg: 0.2 },
  'male:51+': { vitamin_b12_mcg: 0.15, calcium_mg: 0.15 },
};

const HEALTH_FOCUS_WEIGHTS: Record<HealthFocus, Partial<Record<NutrientKey, number>>> = {
  heart: {
    potassium_mg: 0.3,
    magnesium_mg: 0.25,
    fiber_g: 0.2,
    vitamin_b9_mcg: 0.15,
  },
  bone: {
    calcium_mg: 0.35,
    vitamin_k_mcg: 0.25,
    magnesium_mg: 0.2,
    phosphorus_mg: 0.1,
  },
  energy: {
    iron_mg: 0.3,
    vitamin_b12_mcg: 0.25,
    vitamin_b6_mg: 0.2,
    magnesium_mg: 0.15,
    vitamin_b1_mg: 0.15,
  },
  gut: {
    fiber_g: 0.4,
    magnesium_mg: 0.15,
    potassium_mg: 0.1,
  },
  immune: {
    vitamin_c_mg: 0.35,
    zinc_mg: 0.25,
    vitamin_a_mcg: 0.2,
    selenium_mcg: 0.15,
    vitamin_e_mg: 0.1,
  },
};

const PREGNANCY_WEIGHTS: Record<PregnancyStatus, Partial<Record<NutrientKey, number>>> = {
  pregnant: {
    vitamin_b9_mcg: 0.4,
    iron_mg: 0.35,
    calcium_mg: 0.25,
    zinc_mg: 0.2,
  },
  breastfeeding: {
    calcium_mg: 0.3,
    vitamin_b9_mcg: 0.3,
    vitamin_a_mcg: 0.2,
    zinc_mg: 0.2,
  },
};

const LIFESTYLE_WEIGHTS: Record<LifestyleFactor, Partial<Record<NutrientKey, number>>> = {
  smoker: {
    vitamin_c_mg: 0.3,
    vitamin_e_mg: 0.15,
    vitamin_a_mcg: 0.1,
  },
  alcohol: {
    vitamin_b1_mg: 0.25,
    vitamin_b9_mcg: 0.2,
    magnesium_mg: 0.15,
    zinc_mg: 0.1,
  },
  caffeine: {
    iron_mg: 0.2,
    calcium_mg: 0.2,
    magnesium_mg: 0.1,
  },
};

const SYMPTOM_WEIGHTS: Record<Symptom, Partial<Record<NutrientKey, number>>> = {
  fatigue: {
    iron_mg: 0.25,
    vitamin_b12_mcg: 0.25,
    vitamin_b6_mg: 0.15,
    magnesium_mg: 0.1,
  },
  cramps: {
    magnesium_mg: 0.3,
    potassium_mg: 0.25,
    calcium_mg: 0.2,
  },
  bruising: {
    vitamin_c_mg: 0.3,
    vitamin_k_mcg: 0.25,
  },
  colds: {
    vitamin_c_mg: 0.25,
    zinc_mg: 0.2,
    vitamin_a_mcg: 0.15,
    selenium_mcg: 0.1,
  },
};

const FAMILY_HISTORY_WEIGHTS: Record<FamilyCondition, Partial<Record<NutrientKey, number>>> = {
  osteoporosis: {
    calcium_mg: 0.3,
    vitamin_k_mcg: 0.2,
    magnesium_mg: 0.15,
    phosphorus_mg: 0.1,
  },
  heart_disease: {
    potassium_mg: 0.25,
    magnesium_mg: 0.2,
    fiber_g: 0.2,
    vitamin_b9_mcg: 0.15,
  },
  neurodegeneration: {
    vitamin_e_mg: 0.25,
    vitamin_b9_mcg: 0.2,
    vitamin_b12_mcg: 0.2,
    vitamin_b6_mg: 0.15,
  },
  diabetes: {
    fiber_g: 0.3,
    magnesium_mg: 0.25,
    potassium_mg: 0.15,
    zinc_mg: 0.1,
  },
};

const OMEGA3_BONUS_FOODS = new Set([
  'Flaxseed', 'Chia Seeds', 'Walnuts', 'Hemp Seeds',
  'Salmon', 'Sardines', 'Mackerel', 'Trout', 'Herring',
]);

const UNCOMMON_DISCOUNT: Record<string, number> = {
  'Beef Liver': 0.25,
  'Chicken Liver': 0.25,
  'Pork Liver': 0.25,
  'Duck': 0.4,
  'Duck Egg': 0.4,
  'Cornish Hen': 0.4,
  'Pawpaw': 0.4,
  'Teff': 0.4,
  'Amaranth': 0.5,
  'Sorghum': 0.5,
  'Rutabaga': 0.5,
  'Tomatillo': 0.5,
  'Squid': 0.5,
  'Mussel': 0.5,
  'Oyster': 0.5,
  'Herring': 0.6,
  'Anchovy': 0.6,
  'Mackerel': 0.6,
  'Soybean': 0.5,
  'Adzuki Bean': 0.6,
  'Fava Bean': 0.6,
  'Mung Bean': 0.6,
};

const EXCLUDED_NUTRIENTS: Set<NutrientKey> = new Set([
  'vitamin_d_mcg', 'sodium_mg', 'water_g', 'sugars_g',
  'calories_kcal', 'fat_g', 'carbs_g',
]);

function addWeights(
  target: DeficiencyWeights,
  source: Partial<Record<NutrientKey, number>>
): void {
  for (const [key, value] of Object.entries(source)) {
    const k = key as NutrientKey;
    if (EXCLUDED_NUTRIENTS.has(k)) continue;
    target.set(k, Math.min(1, (target.get(k) ?? 0) + value));
  }
}

export function buildDeficiencyProfile(answers: WizardAnswers): DeficiencyWeights {
  const weights: DeficiencyWeights = new Map();

  addWeights(weights, BASELINE_WEIGHTS);
  addWeights(weights, DIET_PATTERN_WEIGHTS[answers.dietPattern]);

  const sexAgeKey = `${answers.sex}:${answers.ageRange}`;
  if (SEX_AGE_WEIGHTS[sexAgeKey]) {
    addWeights(weights, SEX_AGE_WEIGHTS[sexAgeKey]);
  }

  for (const focus of answers.healthFocus) {
    addWeights(weights, HEALTH_FOCUS_WEIGHTS[focus]);
  }

  if (answers.pregnancyStatus) {
    addWeights(weights, PREGNANCY_WEIGHTS[answers.pregnancyStatus]);
  }

  for (const factor of answers.lifestyleFactors) {
    addWeights(weights, LIFESTYLE_WEIGHTS[factor]);
  }

  for (const symptom of answers.symptoms) {
    addWeights(weights, SYMPTOM_WEIGHTS[symptom]);
  }

  for (const condition of answers.familyHistory) {
    addWeights(weights, FAMILY_HISTORY_WEIGHTS[condition]);
  }

  return weights;
}

export function getTopDeficiencies(
  profile: DeficiencyWeights,
  count: number
): { key: NutrientKey; label: string; weight: number }[] {
  const meta = new Map(NUTRIENT_META.map((m) => [m.key, m]));
  return [...profile.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, count)
    .map(([key, weight]) => ({
      key,
      label: meta.get(key)?.label ?? key,
      weight,
    }));
}

const DV_MAP = new Map(
  NUTRIENT_META.filter((m) => m.dailyValue !== null).map((m) => [m.key, m.dailyValue!])
);

const META_MAP = new Map(NUTRIENT_META.map((m) => [m.key, m]));

function getPercentDV(food: NutrientFruit, key: NutrientKey, servingMultiplier: number): number {
  const raw = food[key] as number | null;
  if (raw === null) return 0;
  const dv = DV_MAP.get(key);
  if (!dv) return 0;
  return (raw * servingMultiplier / dv) * 100;
}

export function scoreFoodsForDeficiencies(
  foods: NutrientFruit[],
  profile: DeficiencyWeights,
  maxResults: number = 3
): ScoredFood[] {
  const eligible = foods.filter((f) => f.type !== 'spice' && f.type !== 'fat_oil');

  const scored: { food: NutrientFruit; score: number; nutrients: { key: NutrientKey; pctDV: number }[] }[] = [];

  for (const food of eligible) {
    const servingG = food.serving_size_g ?? 100;
    const multiplier = servingG / 100;
    let score = 0;
    const nutrients: { key: NutrientKey; pctDV: number }[] = [];

    for (const [key, weight] of profile) {
      const pctDV = getPercentDV(food, key, multiplier);
      if (pctDV > 0) {
        score += (pctDV / 100) * weight;
        nutrients.push({ key, pctDV });
      }
    }

    if (OMEGA3_BONUS_FOODS.has(food.name)) {
      score += 0.15;
    }

    const discount = UNCOMMON_DISCOUNT[food.name];
    if (discount !== undefined) {
      score *= discount;
    }

    if (score > 0.05) {
      scored.push({ food, score, nutrients });
    }
  }

  scored.sort((a, b) => b.score - a.score);

  const results: ScoredFood[] = [];
  const usedTypes = new Set<string>();

  for (const item of scored) {
    if (results.length >= maxResults) break;

    let adjustedScore = item.score;
    if (usedTypes.has(item.food.type)) {
      adjustedScore *= 0.6;
    }

    if (adjustedScore < 0.05) continue;

    const topNutrients = item.nutrients
      .sort((a, b) => b.pctDV - a.pctDV)
      .slice(0, 4)
      .map((n) => ({
        key: n.key,
        label: META_MAP.get(n.key)?.label ?? n.key,
        percentDV: Math.round(n.pctDV),
      }));

    results.push({ food: item.food, score: adjustedScore, topNutrients });
    usedTypes.add(item.food.type);
  }

  return results;
}

export function shouldShowB12Note(answers: WizardAnswers): boolean {
  return answers.dietPattern === 'vegan';
}

File: ios/Nutritionmaxx/Nutritionmaxx/Services/DeficiencyProfile.swift
import Foundation

enum BiologicalSex: String, CaseIterable, Codable, Identifiable {
    case male, female
    var id: String { rawValue }
    var label: String {
        switch self {
        case .male: "Male"
        case .female: "Female"
        }
    }
}

enum WizardAgeRange: String, CaseIterable, Codable, Identifiable {
    case young = "19-30"
    case middle = "31-50"
    case senior = "51+"
    var id: String { rawValue }
    var label: String { rawValue }
}

enum DietPattern: String, CaseIterable, Codable, Identifiable {
    case omnivore, pescatarian, vegetarian, vegan
    var id: String { rawValue }
    var label: String { rawValue.capitalized }
}

enum HealthFocus: String, CaseIterable, Codable, Identifiable {
    case heart, bone, energy, gut, immune
    var id: String { rawValue }
    var label: String { rawValue.capitalized }
}

enum PregnancyStatus: String, CaseIterable, Codable, Identifiable {
    case pregnant, breastfeeding
    var id: String { rawValue }
    var label: String { rawValue.capitalized }
}

enum LifestyleFactor: String, CaseIterable, Codable, Identifiable {
    case smoker, alcohol, caffeine
    var id: String { rawValue }
    var label: String { rawValue.capitalized }
}

enum Symptom: String, CaseIterable, Codable, Identifiable {
    case fatigue, cramps, bruising, colds
    var id: String { rawValue }
    var label: String { rawValue.capitalized }
}

enum FamilyCondition: String, CaseIterable, Codable, Identifiable {
    case osteoporosis
    case heartDisease = "heart_disease"
    case neurodegeneration
    case diabetes
    var id: String { rawValue }
    var label: String {
        switch self {
        case .osteoporosis: "Osteoporosis"
        case .heartDisease: "Heart Disease"
        case .neurodegeneration: "Neurodegeneration"
        case .diabetes: "Diabetes"
        }
    }
}

struct WizardAnswers {
    var sex: BiologicalSex = .male
    var ageRange: WizardAgeRange = .young
    var dietPattern: DietPattern = .omnivore
    var healthFocus: Set<HealthFocus> = []
    var pregnancyStatus: PregnancyStatus?
    var lifestyleFactors: Set<LifestyleFactor> = []
    var symptoms: Set<Symptom> = []
    var familyHistory: Set<FamilyCondition> = []
}

struct ScoredFood: Identifiable {
    let food: FoodItem
    let score: Double
    let topNutrients: [(key: NutrientKey, label: String, percentDV: Int)]

    var id: String { food.name }
}

typealias DeficiencyWeights = [NutrientKey: Double]

private let baselineWeights: [NutrientKey: Double] = [
    .fiberG: 0.3, .potassiumMg: 0.3, .magnesiumMg: 0.25,
    .vitaminEMg: 0.2, .vitaminCMg: 0.15, .calciumMg: 0.2, .vitaminAMcg: 0.15,
]

private let dietPatternWeights: [DietPattern: [NutrientKey: Double]] = [
    .omnivore: [:],
    .pescatarian: [.ironMg: 0.15, .zincMg: 0.1],
    .vegetarian: [.vitaminB12Mcg: 0.4, .ironMg: 0.3, .zincMg: 0.2, .calciumMg: 0.15, .proteinG: 0.15, .seleniumMcg: 0.1],
    .vegan: [.vitaminB12Mcg: 0.6, .ironMg: 0.35, .zincMg: 0.3, .calciumMg: 0.3, .proteinG: 0.25, .seleniumMcg: 0.2, .vitaminB2Mg: 0.15],
]

private let sexAgeWeights: [String: [NutrientKey: Double]] = [
    "female:19-30": [.ironMg: 0.3, .vitaminB9Mcg: 0.2],
    "female:31-50": [.ironMg: 0.3, .vitaminB9Mcg: 0.2],
    "female:51+": [.vitaminB12Mcg: 0.2, .calciumMg: 0.2],
    "male:51+": [.vitaminB12Mcg: 0.15, .calciumMg: 0.15],
]

private let healthFocusWeights: [HealthFocus: [NutrientKey: Double]] = [
    .heart: [.potassiumMg: 0.3, .magnesiumMg: 0.25, .fiberG: 0.2, .vitaminB9Mcg: 0.15],
    .bone: [.calciumMg: 0.35, .vitaminKMcg: 0.25, .magnesiumMg: 0.2, .phosphorusMg: 0.1],
    .energy: [.ironMg: 0.3, .vitaminB12Mcg: 0.25, .vitaminB6Mg: 0.2, .magnesiumMg: 0.15, .vitaminB1Mg: 0.15],
    .gut: [.fiberG: 0.4, .magnesiumMg: 0.15, .potassiumMg: 0.1],
    .immune: [.vitaminCMg: 0.35, .zincMg: 0.25, .vitaminAMcg: 0.2, .seleniumMcg: 0.15, .vitaminEMg: 0.1],
]

private let pregnancyWeights: [PregnancyStatus: [NutrientKey: Double]] = [
    .pregnant: [.vitaminB9Mcg: 0.4, .ironMg: 0.35, .calciumMg: 0.25, .zincMg: 0.2],
    .breastfeeding: [.calciumMg: 0.3, .vitaminB9Mcg: 0.3, .vitaminAMcg: 0.2, .zincMg: 0.2],
]

private let lifestyleWeights: [LifestyleFactor: [NutrientKey: Double]] = [
    .smoker: [.vitaminCMg: 0.3, .vitaminEMg: 0.15, .vitaminAMcg: 0.1],
    .alcohol: [.vitaminB1Mg: 0.25, .vitaminB9Mcg: 0.2, .magnesiumMg: 0.15, .zincMg: 0.1],
    .caffeine: [.ironMg: 0.2, .calciumMg: 0.2, .magnesiumMg: 0.1],
]

private let symptomWeights: [Symptom: [NutrientKey: Double]] = [
    .fatigue: [.ironMg: 0.25, .vitaminB12Mcg: 0.25, .vitaminB6Mg: 0.15, .magnesiumMg: 0.1],
    .cramps: [.magnesiumMg: 0.3, .potassiumMg: 0.25, .calciumMg: 0.2],
    .bruising: [.vitaminCMg: 0.3, .vitaminKMcg: 0.25],
    .colds: [.vitaminCMg: 0.25, .zincMg: 0.2, .vitaminAMcg: 0.15, .seleniumMcg: 0.1],
]

private let familyHistoryWeights: [FamilyCondition: [NutrientKey: Double]] = [
    .osteoporosis: [.calciumMg: 0.3, .vitaminKMcg: 0.2, .magnesiumMg: 0.15, .phosphorusMg: 0.1],
    .heartDisease: [.potassiumMg: 0.25, .magnesiumMg: 0.2, .fiberG: 0.2, .vitaminB9Mcg: 0.15],
    .neurodegeneration: [.vitaminEMg: 0.25, .vitaminB9Mcg: 0.2, .vitaminB12Mcg: 0.2, .vitaminB6Mg: 0.15],
    .diabetes: [.fiberG: 0.3, .magnesiumMg: 0.25, .potassiumMg: 0.15, .zincMg: 0.1],
]

private let omega3BonusFoods: Set<String> = [
    "Flaxseed", "Chia Seeds", "Walnuts", "Hemp Seeds",
    "Salmon", "Sardines", "Mackerel", "Trout", "Herring",
]

private let uncommonDiscount: [String: Double] = [
    "Beef Liver": 0.001, "Chicken Liver": 0.001, "Pork Liver": 0.001,
    "Duck": 0.4, "Duck Egg": 0.4, "Cornish Hen": 0.1, "Pawpaw": 0.2,
    "Teff": 0.2, "Pork Belly": 0.3, "Amaranth": 0.2, "Sorghum": 0.2,
    "Rutabaga": 0.2, "Tomatillo": 0.5, "Squid": 0.5, "Mussel": 0.5,
    "Oyster": 0.5, "Soybean": 0.5, "Millet": 0.5, "Buckwheat": 0.5,
    "Durum Wheat": 0.5, "Spelt": 0.5, "Clam": 0.5, "Scallop": 0.5,
    "Lobster": 0.5, "Great Northern Bean": 0.5, "Navy Bean": 0.5,
    "Split Pea": 0.5, "Herring": 0.2, "Anchovy": 0.5, "Mackerel": 0.5,
    "Adzuki Bean": 0.3, "Fava Bean": 0.3, "Mung Bean": 0.3,
    "Collard Greens": 0.8, "Parsnip": 0.6, "Fennel": 0.6,
    "Bok Choy": 0.6, "Crab (King)": 0.6, "Wild Rice": 0.6,
]

private let excludedDeficiencyNutrients: Set<NutrientKey> = [
    .vitaminDMcg, .sodiumMg, .waterG, .sugarsG, .caloriesKcal, .fatG, .carbsG,
]

private func addWeights(_ target: inout DeficiencyWeights, _ source: [NutrientKey: Double]) {
    for (key, value) in source {
        if excludedDeficiencyNutrients.contains(key) { continue }
        target[key] = min(1, (target[key] ?? 0) + value)
    }
}

func buildDeficiencyProfile(_ answers: WizardAnswers) -> DeficiencyWeights {
    var weights: DeficiencyWeights = [:]
    addWeights(&weights, baselineWeights)
    addWeights(&weights, dietPatternWeights[answers.dietPattern] ?? [:])

    let sexAgeKey = "\(answers.sex.rawValue):\(answers.ageRange.rawValue)"
    if let w = sexAgeWeights[sexAgeKey] {
        addWeights(&weights, w)
    }

    for focus in answers.healthFocus {
        addWeights(&weights, healthFocusWeights[focus] ?? [:])
    }

    if let status = answers.pregnancyStatus {
        addWeights(&weights, pregnancyWeights[status] ?? [:])
    }

    for factor in answers.lifestyleFactors {
        addWeights(&weights, lifestyleWeights[factor] ?? [:])
    }

    for symptom in answers.symptoms {
        addWeights(&weights, symptomWeights[symptom] ?? [:])
    }

    for condition in answers.familyHistory {
        addWeights(&weights, familyHistoryWeights[condition] ?? [:])
    }

    return weights
}

func getTopDeficiencies(_ profile: DeficiencyWeights, count: Int) -> [(key: NutrientKey, label: String, weight: Double)] {
    profile.sorted { $0.value > $1.value }
        .prefix(count)
        .map { (key: $0.key, label: nutrientMetaMap[$0.key]?.label ?? $0.key.rawValue, weight: $0.value) }
}

private let dvMap = Dictionary(uniqueKeysWithValues: nutrientMetaList.compactMap { m -> (NutrientKey, Double)? in
    guard let dv = m.dailyValue else { return nil }
    return (m.key, dv)
})

private func getPercentDV(_ food: FoodItem, _ key: NutrientKey, _ servingMultiplier: Double) -> Double {
    guard let raw = food[key], let dv = dvMap[key], dv > 0 else { return 0 }
    return (raw * servingMultiplier / dv) * 100
}

private func scoreFood(_ food: FoodItem, _ weights: DeficiencyWeights) -> (score: Double, nutrients: [(key: NutrientKey, pctDV: Double)]) {
    let multiplier = food.servingMultiplier
    var score = 0.0
    var nutrients: [(key: NutrientKey, pctDV: Double)] = []

    for (key, weight) in weights {
        let pctDV = getPercentDV(food, key, multiplier)
        if pctDV > 0 {
            score += (pctDV / 100) * weight
            nutrients.append((key: key, pctDV: pctDV))
        }
    }

    if omega3BonusFoods.contains(food.name) {
        score += 0.15
    }

    if let discount = uncommonDiscount[food.name] {
        score *= discount
    }

    return (score, nutrients)
}

func scoreFoodsForDeficiencies(_ foods: [FoodItem], _ profile: DeficiencyWeights, maxResults: Int = 3) -> [ScoredFood] {
    let eligible = foods.filter { $0.type != .spice && $0.type != .fatOil }
    var remainingWeights = profile
    var results: [ScoredFood] = []
    var usedTypes: Set<ItemType> = []
    var usedFoods: Set<String> = []

    for _ in 0..<maxResults {
        var best: (food: FoodItem, score: Double, nutrients: [(key: NutrientKey, pctDV: Double)])?

        for food in eligible {
            if usedFoods.contains(food.name) { continue }
            var (score, nutrients) = scoreFood(food, remainingWeights)
            if usedTypes.contains(food.type) {
                score *= 0.6
            }
            if score > 0.05 && (best == nil || score > best!.score) {
                best = (food, score, nutrients)
            }
        }

        guard let chosen = best else { break }

        let topNutrients = chosen.nutrients
            .sorted { $0.pctDV > $1.pctDV }
            .prefix(4)
            .map { (key: $0.key, label: nutrientMetaMap[$0.key]?.label ?? $0.key.rawValue, percentDV: Int($0.pctDV.rounded())) }

        results.append(ScoredFood(food: chosen.food, score: chosen.score, topNutrients: topNutrients))
        usedFoods.insert(chosen.food.name)
        usedTypes.insert(chosen.food.type)

        for (key, pctDV) in chosen.nutrients {
            guard let currentWeight = remainingWeights[key] else { continue }
            let coverage = min(pctDV / 100, 1)
            let reduction = coverage * 0.8
            let newWeight = currentWeight * (1 - reduction)
            if newWeight < 0.01 {
                remainingWeights.removeValue(forKey: key)
            } else {
                remainingWeights[key] = newWeight
            }
        }
    }

    return results
}
